--- a/types/wlr_linux_dmabuf_v1.c.orig
+++ b/types/wlr_linux_dmabuf_v1.c
@@ -324,8 +324,7 @@
 			goto err_out;
 		}
 
-		if (attribs.offset[i] + attribs.stride[i] > size ||
-				attribs.stride[i] == 0) {
+		if (attribs.offset[i] + attribs.stride[i] > size) {
 			wl_resource_post_error(params_resource,
 				ZWP_LINUX_BUFFER_PARAMS_V1_ERROR_OUT_OF_BOUNDS,
 				"invalid stride %" PRIu32 " for plane %d",
