--- a/include/linux/mfd/rk808.h	2026-03-19 10:13:45.918393258 +1300
+++ b/include/linux/mfd/rk808.h	2026-03-19 10:16:52.640926235 +1300
@@ -1466,6 +1466,11 @@
 	long				variant;
 	const struct regmap_config	*regmap_cfg;
 	const struct regmap_irq_chip	*regmap_irq_chip;
+
+	/* Lazy-loaded pinctrl for RK817/RK809 SLPPIN poweroff */
+	bool				pctrl_loaded;
+	struct pinctrl			*pctrl;
+	struct pinctrl_state		*pins_pwroff;
 };
 
 void rk8xx_shutdown(struct device *dev);
